plugin/emacs/flycheck-quicklintjs.el

```diff
@@ -0,0 +1,101 @@
+;;; flycheck-quicklintjs --- Quick Lint JS flycheck support   -*- lexical-binding: t; -*-
+
+;;; Commentary:
+
+;; This flycheck exetension configures flycheck to run Quick Lint JS
+;; to run as a checker.
+;;
+;; Example usage in your init.el:
+;;
+;; (require 'flycheck-quicklintjs)
+;;
+;; ;; Optionally you can setup flycheck with the expected
+;; ;; quick lint js experience:
+;; (defun my-flycheck-quicklintjs-setup ()
+;;   "Configure flycheck-quicklintjs for better experience."
+;;
+;;   ;; Enable flycheck
+;;   (unless (bound-and-true-p flycheck-mode)
+;;     (flycheck-mode))
+;;
+;;   ;; Use quick-lint-js by default when in 'js-mode`
+;;   (flycheck-select-checker 'javascript-quicklintjs)
+;;
+;;   ;; Remove any delay after a change in buffer to run checkers.
+;;   ;; The default is 0.5 (500ms)
+;;   (setq-local flycheck-idle-change-delay 0)
+;;
+;;   ;; Arguments to quick-lint-js program
+;;   ;; (setq-local flycheck-quicklintjs-args "--output-format=gnu-like")
+;;
+;;   ;; Change the quick-lint-js program to run.
+;;   ;; If not set, Emacs executable-find function is used, which searches in
+;;   ;; exec-path variable, probably you don't need to set this if quick-lint-js
+;;   ;; is installed in your system.
+;;   ;; (setq-local flycheck-javascript-quicklintjs-executable "~/quick-lint-js")
+;;
+;;   ;; Run quick-lint-js program when the buffer is changed and when 'js-mode` is
+;;   ;; loaded
+;;   (setq-local flycheck-check-syntax-automatically '(mode-enabled idle-change)));
+;; (add-hook 'js-mode-hook #'my-flycheck-quicklintjs-setup)
+;;
+
+;;; Code:
+
+(require 'flycheck)
+
+(defgroup flycheck-quicklintjs nil
+  "Quick Lint JS Flycheck integration."
+  :prefix "flycheck-"
+  :group 'flycheck
+  :group 'quicklintjs
+  :link '(url-link :tag "Website" "https://quick-lint-js.com"))
+
+(flycheck-def-args-var flycheck-quicklintjs-args javascript-quicklintjs)
+(flycheck-def-executable-var javascript-quicklintjs "quick-lint-js")
```

Comment:
If quick-lint-js is installed locally in a project via npm, ideally we should use the npm-installed version instead of the globally-installed version. (This would mean that some buffers might use the npm-installed version and others might not. So maybe this kind of setup isn't possible with flycheck.)

Vim's ALE has a mechanism to dynamically search for the linter per-buffer. Does Flycheck have a similar mechanism? If not, I think we should mention this limitation in the documentation. Specifically, something like "If you installed quick-lint-js in your project using npm, you need to manually configure the Emacs plugin to use your project's version. Here's how: ...".

Reply:
I see, but in fact this isn't how we define the program to run, we define below as a literal string "quick-lint-js",  in fact this is a macro which its expansion will define flycheck-SYMBOL-executable, in this case flycheck-javascript-quicklintjs-executable, it's a variable meant for the user to change, not us, it's here so users inspect what the default value is with emacs builtin help system (C-h v).

Flycheck is a emacs minor mode, as far as im concern, every minor mode is **buffer-local**, meaning, it's possible to change dynamically which linter to run, the problem here is that emacs as vim is a text editor, it's doesnt have a notion of a "project", just open text files, some emacs users use a popular package named projectile for that.  But i see that we can change :command to be a generic lisp function, which traverses back the parent directories until it founds a node_modules with quick-lint-js? im not sure if it's what general users expect, but im sure it's what a npm user would expect.

Reply:
>  But i see that we can change :command to be a generic lisp function, which traverses back the parent directories until it founds a node_modules with quick-lint-js? im not sure if it's what general users expect, but im sure it's what a npm user would expect.

That solution makes sense to me.

What does ESLint's Emacs plugin do?

Reply:
@strager here's what flycheck.el does for eslint:
```
  :command ("eslint" "--format=json"
            (option-list "--rulesdir" flycheck-eslint-rules-directories) ;; <- a customizablie list of directories that contains eslint rules, what that means.
            (eval flycheck-eslint-args)
            "--stdin" "--stdin-filename" source-original)
  :enabled (lambda () (flycheck-eslint-config-exists-p)) ;; <- checks for .eslintrc, apparently
```
it doesnt look like to make anything more than that, probably it's leftover for users to overwrite this with a js-mode hook that does that.

Reply:
self correcting me:
> it's doesnt have a notion of a "project"

I just learned from the survey i linked below about the builtin "package.el", it was introduced in emacs 26.1, i'll need investigate that if it's of any value for us.